### PR TITLE
Make global styles data filterable

### PR DESCRIPTION
**Pull request by oandregal**
## What?

This PR tries making the different layers of global styles data filterable.

## Why?

Some use cases become a lot easier to do:

- Some themes may want to provide settings dynamically. For example, color presets depending on some contextual information.
- Allow plugins to modify the existing data. For example, provide defaults for many blocks at once. See #41707
- Some agencies may want to edit or remove core data (color presets, etc).

## How?

By creating a filter for each data layer of Global Styles that 3rd parties can use:

- `global_styles_default` => to filter data provided by core
- `global_styles_blocks` => to filter data provided by the blocks (only styles so far)
- `global_styles_theme` => to filter data provided by the theme
- `global_styles_user` => to filter data provided by the user

## Prior work

There were previous attempts to add filters to global styles that didn't make the cut for different reasons.

- Filters to modify the stylesheet and the low-level schema https://github.com/WordPress/gutenberg/pull/25499 by @aristath 
- Filter the stylesheet and theme support data https://github.com/WordPress/gutenberg/pull/27509 by @aristath 
- Filters to modify the low-level schema https://github.com/WordPress/gutenberg/pull/27722 by @oandregal 
- Filter the stylesheet generated by global styles https://github.com/WordPress/gutenberg/pull/28687 by @aristath 
- Make the theme's theme.json extensible https://github.com/WordPress/gutenberg/pull/30259 by @scruffian 
- Filters to shortcircuit the public API (`wp_get_global_*` functions) https://github.com/WordPress/gutenberg/pull/36909 by @oandregal

## Testing Instructions

Use the filters to modify some of the data. For example, this is how the theme data can be modified:

```php
<?php

function filter_global_styles_theme( $theme_json ){
        $new_data = array(
                'version'  => 2,
                'settings' => array(
                        'color' => array(
                                'text'  => false, /* disable text color UI */
                                'palette'    => array( /* New palette */
                                        array(
                                                'slug'  => 'foreground',
                                                'color' => 'red',
                                                'name'  => 'Foreground',
                                        ),
                                ),
                        ),
                ),
        );
        return $theme_json->update_with( $new_data );
}
add_filter( 'global_styles_theme', 'filter_global_styles_theme' );
```

## TODO / Follow-ups

- How to make `set_spacing_sizes` filterable as well. See https://github.com/WordPress/gutenberg/pull/41527/files#r966773252
- ~The changes for the resolver live in `lib/experimental` though some are expected to be ported to 6.1.~ The changes that affect this PR have been ported.

### CODE REVIEW

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php
Comment (aristath) [inline lib/experimental/class-wp-theme-json-resolver-gutenberg.php]:
```suggestion
			$result->merge( apply_filters( 'global_styles_user', static::get_user_data() ) );
```

**Review by aristath — APPROVED** [review]
YES!! 
I've tried adding filters multiple times but it was too soon for those PRs. I think now we're at a point in time where we can do this with confidence.
I only had one note for this PR, and this is the addition of a `global_styles_user` filter 👍

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php
Comment (oandregal) [reply lib/experimental/class-wp-theme-json-resolver-gutenberg.php]: I'm not sure about this one. Is there any precedent in core for filtering user data? In principle I'd think that user data shouldn't be modified by anyone.

Comment (aristath) [reply lib/experimental/class-wp-theme-json-resolver-gutenberg.php]: The reason I'm suggesting this one, is that it's at the "end of the line" - where we'd expect to see a filter. Most WP filters are at the end of functions and filter the return values - including user data.
Ideally I'd like to see a `global_styles` filter at the `return` of the function as that would provide the maximum amount of flexibility without the need for separate filters per-context.
But if we do add separate filters for each item, then user styles will override any filters added - which would "neutralize" them. That may be the intention since user data should probably override the others, however it gives the impression that the filters are "broken" since they won't work.
Perhaps we could add a `global_styles` filter at the return of this function, and pass the core/blocks/theme/user data as params to the filter?

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php
Comment (oandregal) [inline lib/experimental/class-wp-theme-json-resolver-gutenberg.php]: Just in the line below there's `set_spacing_sizes`, a configuration for spacing presets that does not follow the normal algorith (core > theme > user), so it wouldn't be filterable. I need to figure out how that works and whether it can work like any other `theme.json` setting. See https://github.com/WordPress/gutenberg/pull/41527/files#r966773252

**Comment by oandregal** [comment]
With the current proposal, this is what a theme should do to provide dynamic values:

```php
<?php

function filter_global_styles_theme( $theme_json ){
	$dynamic_data = array(
		'version'  => 2,
		'settings' => array(
			'color' => array(
				'text'       => false, /* New setting */
				'palette'    => array( /* New palette */
					array(
						'slug'  => 'foreground',
						'color' => 'red',
						'name'  => 'Foreground',
					),
					array(
						'slug'  => 'New foreground',
						'color' => 'hotpink',
						'name'  => 'New foreground',
					),
				),
			),
		),
	);

	$theme_json->merge( new WP_Theme_JSON( $dynamic_data, 'theme' ) );
	return $theme_json;
}
add_filter( 'global_styles_theme', 'filter_global_styles_theme' );
```

A few things here:

- The author needs to provide a version. I think we should maintain this requirement, to prevent people using this hook from having to update when there's a new `theme.json` version (changes to shape, etc). We already provide migrations at runtime for this, so we should use them here as well.
- The ergonomics can be improved. We still need to make sure this data goes through the following processes: migration, validation, origin management, merging. In this example I use `WP_Theme_JSON` class, which has been kept private so far. I think we can provide a different public API that complements [the existing](https://make.wordpress.org/core/2022/01/04/new-api-to-access-global-settings-styles/).

**Comment by oandregal** [comment]
Pushed a few changes, and now consumers can use the new `gutenberg_merge_theme_json` to do the heavy lifting for them:

(edit: pushed https://github.com/WordPress/gutenberg/pull/44015/commits/5d3b6380d74440ec9a10cf316b400bc0d035fbee so `$origin` is optional in `gutenberg_merge_theme_json`):

```php
<?php

function filter_global_styles_theme( $existing ){
	$dynamic_data = array(
		'version'  => 2,
		'settings' => array(
			'color' => array(
				'text'       => false, /* New setting */
				'palette'    => array( /* New palette */
					array(
						'slug'  => 'foreground',
						'color' => 'red',
						'name'  => 'Foreground',
					),
					array(
						'slug'  => 'New foreground',
						'color' => 'hotpink',
						'name'  => 'New foreground',
					),
				),
			),
		),
	);
	$filtered = gutenberg_merge_theme_json( $existing, $dynamic_data );

	return $filtered;
}
add_filter( 'global_styles_theme', 'filter_global_styles_theme' );
```

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php
Comment (oandregal) [reply lib/experimental/class-wp-theme-json-resolver-gutenberg.php]: For user data: according to [how the API is evolving](https://github.com/WordPress/gutenberg/pull/44015#issuecomment-1241899708), adding a `global_styles_user` hook is a no-brainer. Though why someone would want to filter user styles and presets?

For the merged data: if we do this it'll be more involved for consumers to filter. At this point we've already merged the data coming from different origins (defaults, theme, and user), so consumers would need to have some understanding of what this means — for example, how we store the palettes for different origins. I'm less convinced about this one. It requires less knowledge from consumers if they just filter the bits they want to modify.

File: lib/compat/wordpress-6.1/merge-theme-json.php
Comment (oandregal) [inline lib/compat/wordpress-6.1/merge-theme-json.php]: By filling the `$origin` depending on which filter is running, we avoid asking consumers to do it themselves. Though if the consumer does provide it, we use it instead.

**Comment by mtias** [comment]
As mentioned in Slack, absolutely in favor of adding this. Low level hooks at all the levels in which we construct global style objects is necessary and should provide flexibility for plugins and sites to do a lot of creative things.

File: lib/compat/wordpress-6.1/merge-theme-json.php
Comment (nerrad) [reply lib/compat/wordpress-6.1/merge-theme-json.php]: I like this. It abstracts away this requirement which might not be easily understood by consumers implementing the filter.

**Review by nerrad — COMMENTED** [review]
I think the direction of this is good. I do wonder if there should be some validation of returned filter values. While I appreciate the availability of `gutenberg_merge_them_json` to handle merging, it's not something intuitive for consumers to implement on whatever they are merging into existing objects. There's also increased fragility. I wonder if it'd be better to pass through an instance of a `WP_Theme_JSON` object and consumers would be expected to use the `merge` method and then _return_ an instance of `WP_Theme_JSON` (which could be validated).

So for instance the theme filter would be something like this:

```php
<?php
$unfiltered_json_data = new WP_Theme_JSON_Gutenberg( $theme_json_data );
try {
  $theme_json_data = apply_filters( 'global_styles_theme', $unfiltered_json_data );
  if ( ! $theme_json_data instanceof WP_Theme_JSON ) {
    throw new Exception( 'Incorrect usage of filter' );
  }
} catch (Exception $e) {
  // just handling incorrect usage of filter and resetting to unfiltered data
  $theme_json_data = $unfiltered_json_data;
}

static::$theme   = $theme_json_data;
```

And implementation would be:

```php
function filter_global_styles_theme( WP_Theme_JSON $theme_json ){
	$dynamic_data = array(
		'version'  => 2,
		'settings' => array(
			'color' => array(
				'text'       => false, /* New setting */
				'palette'    => array( /* New palette */
					array(
						'slug'  => 'foreground',
						'color' => 'red',
						'name'  => 'Foreground',
					),
					array(
						'slug'  => 'New foreground',
						'color' => 'hotpink',
						'name'  => 'New foreground',
					),
				),
			),
		),
	);

  /**
   * This would be a new method on WP_Theme_JSON for merging in raw data.
   * It internally handles converting the dynamic data to a WP_Theme_JSON object
   * and the consumer would not need to provide $origin.
   */
	$theme_json->merge_raw( $dynamic_data );
	return $theme_json;
}
add_filter( 'global_styles_theme', 'filter_global_styles_theme' );
```

I think this goes a bit back towards your original solution with the change of implementing a new method on `WP_Theme_JSON` for handling merging raw data.

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php
Comment (nerrad) [reply lib/experimental/class-wp-theme-json-resolver-gutenberg.php]: I don't think user styles should be filterable. The use-case I can think of for that filter would be as a fail-safe for users making style/preset changes that a design system might not be allowing. However, that's probably something that should be controllable via `supports` properties?

**Comment by oandregal** [comment]
> I think the direction of this is good. I do wonder if there should be some validation of returned filter values. 

The returned values will be validated in the same way a `theme.json` provided by the theme is. Note that after the filters, the next call is `WP_Theme_JSON( $filtered_data )`.

**Comment by oandregal** [comment]
> I wonder if it'd be better to pass through an instance of a WP_Theme_JSON object and consumers would be expected to use the merge method and then return an instance of WP_Theme_JSON (which could be validated).

Having tested both approaches, using a simple array is more familiar and ergonomic for consumers: it's a construct people know how to work with in the language, and it mimics the mechanics of `array_merge` but it's tailored to `theme.json` data.

At first sight, the two different approaches may look similar:

```php
$existing->merge_raw( $new_data );
```

vs 

```php
gutenberg_merge_theme_json( $existing, $new_data );
```

Though the 1st approach forces us to making `WP_Theme_JSON` public (exposing a  lot of unnecessary complexity to consumers) and it's more inflexible (we can't get rid of the class in the future, should we want).

**Comment by oandregal** [comment]
Note for future readers: while editing directly the given array is doable (e.g.: set to `false` the value of `settings.color.text`) I don't recommend it as it's not future-proof. If a future version of `theme.json` updates the text color flag to be in a different place, the consumers that do this will have to update their code. By using the helper function they won't have to do anything.

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php
Comment (oandregal) [reply lib/experimental/class-wp-theme-json-resolver-gutenberg.php]: Added the filter for `user` but not for `merged`.

**Comment by oandregal** [comment]
This is ready for review. If we want it in 6.1 it should land before tomorrow's RC.

There're two things that can be follow-ups: how changes in `lib/experimental` are backported to core and whether we need to modify how `set_spacing_sizes` works for making them filterable.

**Comment by oandregal** [comment]
@aristath I know you already approved when this PR was merely a draft and not fully working/developed. I'd welcome a technical review now if you've got time.

**Comment by gziolo** [comment]
So great to see filters added to the global styles 🎉 

I have a question regarding the filters that folks will write. Is it something you would recommend to directly manipulate `$existing`?

```php
<?php

function filter_global_styles_theme( $existing ){
    $existing['settings']['text'] = false;
    $existing['settings']['color']['pallete'][] = array(
        'slug'  => 'foreground',
        'color' => 'red',
        'name'  => 'Foreground',
    );
    return $existing;
}
add_filter( 'global_styles_theme', 'filter_global_styles_theme' );
```

**Comment by nerrad** [comment]
>  Note for future readers: while editing directly the given array is doable (e.g.: set to false the value of settings.color.text) I don't recommend it as it's not future-proof. If a future version of theme.json updates the text color flag to be in a different place, the consumers that do this will have to update their code. By using the helper function they won't have to do anything.

This is my single concern about the current approach. If consumers do not utilize `gutenberg_merge_theme_json` and simply modify the given array, things could work and there is no warning that this could break in the future (and unexpectedly). 

If this concern isn't shared by other folks I'm happy to see this merge as is - it'll be a good solution for some use-cases.

**Comment by oandregal** [comment]
> I have a question regarding the filters that folks will write. Is it something you would recommend to directly manipulate $existing?

This was my earlier attempt to respond https://github.com/WordPress/gutenberg/pull/44015#issuecomment-1243550047 (I didn't share code snippets intentionally :sweat_smile: ).

I don't recommend doing that unless the consumer is on top of the theme.json changes and can update their code. It'll work and probably it's going to be fine most of the time, though the issue is this: if, in the future, we need to update the `theme.json` shape (for example, moving `settings.color.text` to `settings.textColor`) the consumers that use the direct manipulaction will need to update their code. This is an unlikely situation but we should be prepared anyway.

**Comment by gziolo** [comment]
In that case, one way to approach that would be putting `$existing` behind the tiny class wrapper that implements [`ArrayAccess`](https://www.php.net/manual/en/class.arrayaccess.php) with a lock for direct modifications:

```php
class WP_Filter_Theme_JSON implements ArrayAccess {
    private $existing = array();

    public function __construct( $existing ) {
        $this->existing = $existing;
    }

    public function offsetSet( $offset, $value ) {
        // Not implemented to prevent direct modifications.
    }

    public function offsetExists( $offset ) {
        return isset( $this->existing[ $offset ] );
    }

    public function offsetUnset( $offset ) {
        // Not implemented to prevent direct modifications.
    }

    public function offsetGet( $offset ) {
        return isset($this-> existing[$offset]) ? $this-> existing[$offset] : null;
    }

    public function merge( $new_data ) {
        return gutenberg_merge_theme_json( $this->existing, $new_data );
    }
}
```

It probably might require more work to prevent modifications for nested items. 

Anyway, with that in place folks would have use the following:

```php
<?php

function filter_global_styles_theme( $existing ){
	$dynamic_data = array(
		'version'  => 2,
		'settings' => array(
			'color' => array(
				'text'       => false, 
				'palette'    => array( 
					array(
						'slug'  => 'foreground',
						'color' => 'red',
						'name'  => 'Foreground',
					),
				),
			),
		),
	);
	return $existing->merge( $dynamic_data );
}
add_filter( 'global_styles_theme', 'filter_global_styles_theme' );
```

**Comment by oandregal** [comment]
In https://github.com/WordPress/gutenberg/pull/44015/commits/c340a37dbcfce4fdfcf13c950fa0ac3b2fe74feb I've updated the code so the filters don't use the `theme.json` array directly but an instance of a new class `WP_Theme_JSON_Data`. Consumers have to do this to filter the `theme.json`:

```php
<?php

function filter_global_styles_theme( $theme_json ){
        $new_data = array(
                'version'  => 2,
                'settings' => array(
                        'color' => array(
                                'text'  => false, /* disable text color UI */
                                'palette'    => array( /* New palette */
                                        array(
                                                'slug'  => 'foreground',
                                                'color' => 'red',
                                                'name'  => 'Foreground',
                                        ),
                                ),
                        ),
                ),
        );
        return $theme_json->update_with( $new_data );
}
add_filter( 'global_styles_theme', 'filter_global_styles_theme' );
```

File: lib/compat/wordpress-6.1/class-wp-theme-json-data.php
Comment (nerrad) [inline lib/compat/wordpress-6.1/class-wp-theme-json-data.php:25]: This should probably default to `''` (empty string) if the type is explicitly set as string.

File: lib/compat/wordpress-6.1/class-wp-theme-json-data.php
Comment (nerrad) [inline lib/compat/wordpress-6.1/class-wp-theme-json-data.php:47]: Unnecessary line break here?

**Review by nerrad — APPROVED** [review]
I definitely like the addition of `WP_Theme_JSON_Data` which:

- addresses ergonomic usage of the filters by consumers while ensuring proper use.
- helps keep the implementation of `WP_Theme_JSON` internal (allowing for future changes without impacting filter usage).

I have a few minor nitpicks in comments and am fine merging as is, however ideally it'd be good to see some simple PHPUnit tests covering filter usage testing expected behaviour as well as improper incoming array format (unless that's already covered in `WP_Theme_JSON` tests).

**Comment by oandregal** [comment]
Thanks for the feedback, folks. I'm going to merge as it is to expedit it for today's RC and will keep an eye on follow-ups.

@nerrad yeah, the validation is already covered by `WP_Theme_JSON` [tests](https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/tests/theme/wpThemeJson.php).

File: lib/compat/wordpress-6.1/class-wp-theme-json-data.php
Comment (andrewserong) [inline lib/compat/wordpress-6.1/class-wp-theme-json-data.php:11]: Apologies for commenting just after this PR merged, but should this line have a `Gutenberg` suffix, or have a `class_exists` check before defining it, so that we don't run into any issues when this is backported to core?

**Comment by aristath** [comment]
> @aristath I know you already approved when this PR was merely a draft and not fully working/developed. I'd welcome a technical review now if you've got time.

I'm a bit late since this was already merged, but I'll leave my feedback anyway; this implementation makes sense and looks good to me 👍 
Thank you!

File: lib/compat/wordpress-6.1/class-wp-theme-json-data.php
Comment (gziolo) [reply lib/compat/wordpress-6.1/class-wp-theme-json-data.php:11]: Yes, it'd be best to call `class_exists` before requiring this file.

Comment (aristath) [reply lib/compat/wordpress-6.1/class-wp-theme-json-data.php:11]: +1

**Comment by gziolo** [comment]
Great job bringing this to the finish line @oandregal. The way you implemented it is even better than what I initially drafted in https://github.com/WordPress/gutenberg/pull/44015#issuecomment-1243906395, as the new class might be applicable in different contexts, too. It might seem like overhead adding a wrapper class at first glance, but it also gives you space for future optimizations as you have control over the input and output value of the filter.

As a follow-up, we need to document all new filters inline in the codebase like:

https://github.com/WordPress/wordpress-develop/blob/1746a68584dbadb8807a1520546725f8a1e572df/src/wp-includes/class-wp-block-type.php#L381-L389

This is going to be necessary when backporting to WordPress core. 

In addition to that, we could extend documentation in the block editor handbook. It might even be worth adding a new document specific to themes in:

https://github.com/WordPress/gutenberg/tree/trunk/docs/reference-guides/filters

File: lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php
Comment (gziolo) [inline lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php:60]: With WP filters we need to return the same data type as in the input, so that should be `$data` instead of config, so maybe the following would be more accurate:
```php
$data = apply_filters( 'global_styles_default', new WP_Theme_JSON_Data( $config, 'default' ) );
static::$core = new WP_Theme_JSON_Gutenberg( $data->get_config(), 'default' );
```

I see the same pattern in other places. `WP_Theme_JSON` lists `$theme_json` in the constructor:

https://github.com/WordPress/wordpress-develop/blob/1746a68584dbadb8807a1520546725f8a1e572df/src/wp-includes/class-wp-theme-json.php#L431-L440

Maybe, we could rename that to `$theme_config` in WP core and here it would be `$data->get_theme_config()`?

**Comment by oandregal** [comment]
@nerrad @andrewserong @gziolo Addressed your feedback at https://github.com/WordPress/gutenberg/pull/44109

**Comment by oandregal** [comment]
https://github.com/WordPress/gutenberg/pull/44111 documents the new filters in the block editor handbook

**Comment by fabiankaegy** [comment]
I've added the Needs Dev Note label here because I think this should get a dev note for the 6.1 release.

CC: @bph

**Comment by oandregal** [comment]
The devnote content can actually be the same as the [documentation](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/filters/global-styles-filters.md). It can work either as a standalone note or as part of a larger one. Would we publish a consolidated devnote for all global styles work in WordPress 6.1? (I've been out a few months in this cycle and don't have a good context of what has changed).

**Comment by oandregal** [comment]
https://github.com/WordPress/gutenberg/pull/44159 proposes we rename the filters from `global_styles_*` to `theme_json_*`.

**Comment by oandregal** [comment]
This is being backported as part of https://github.com/WordPress/wordpress-develop/pull/3247

**Comment by bph** [comment]
> Would we publish a consolidated devnote for all global styles work in WordPress 6.1

@oandregal Yes, I think you should. I am working through collected all the PRs but I am uncertain whether I get them all.  😸

**Comment by adamziel** [comment]
I like how this PR opens the door for updating the global styles in plugins without introducing new core APIs.

That being said, I’m a bit worried, too. Imagine 5 different plugins competing to update similar config values. I can't think of a clean way of solving it, but if this was a declarative API then it could at least run some checks and notify developers about, say, conflicting updates.

Perhaps I’m thinking of `theme.json` dev tools with a traceable cascade of values, like what Chrome has for CSS. That's a long way to go, though. Perhaps introducing a filter today is for the best after all. Implementing it as `$theme_json->update_with( $new_data );` is perfect because it makes the source of the changes traceable.

**Pull request merged**
